GeomIO/Grid_PFIO.F90

```diff
@@ -0,0 +1,63 @@
+#include "MAPL_Generic.h"
+
+module mapl3g_GridPFIO
+   use mapl_ErrorHandling
+   use mapl3g_GeomPFIO
+   use ESMF
+   use PFIO
+   use MAPL_BaseMod
+   implicit none
+   private
+
+   public :: GridPFIO
+   type, extends (GeomPFIO) :: GridPFIO
+      private
+   contains
+      procedure :: stage_data_to_file
+   end type GridPFIO
+
+
+contains
+
+   subroutine stage_data_to_file(this, bundle, filename, time_index, rc)
+      class(GridPFIO), intent(inout) :: this
+      type(ESMF_FieldBundle), intent(in) :: bundle
+      character(len=*), intent(in) :: filename
+      integer, intent(in) :: time_index
+      integer, intent(out), optional :: rc
+
+      integer :: status, num_fields, i, collection_id
+      character(len=ESMF_MAXSTR), allocatable :: field_names(:)
+      type(ESMF_Field) :: field
+      type(ArrayReference) :: ref
+      real, pointer :: ptr2d(:,:)
+      integer, allocatable :: local_start(:), global_start(:), global_count(:)
+
+      type(ESMF_Grid) :: grid
+      integer :: global_dim(3), i1, j1, in, jn
+
+      collection_id = this%get_collection_id()
+      call ESMF_FieldBundleGet(bundle, fieldCount=num_fields, _RC)
+      allocate(field_names(num_fields))
+      call ESMF_FieldBundleGet(bundle, fieldNameList=field_names, _RC)
+      do i=1,num_fields
+         call ESMF_FieldBundleGet(bundle, field_names(i), field=field, _RC)
+         ! all this logic needs to be generalized
+         call ESMF_FieldGet(field, farrayPtr=ptr2d, _RC)
```

Comment:
I think @darianboggs has a procedure that will return  a 1D array ptr regardless of actual rank of the field.   This should avoid conditional logic on the rank.   But more thought is needed about avoiding conditional logic on type (R4, R8, I4, I8, L).

Reply:
@tclune, yes this needs thoughts, however we also have the issue that for certain grid classes (aka cube the pointer we need to pass the array reference gets that extra dimensions, hence the proliferation of methods in the old grid factory...

Reply:
The idea is that you would separately determine shape (and offsets).   But not add any conditional logic on the rank/type of the Field pointer itself.    I think I see how that will work well for rank.  But type is more problematic, or at least requires a possibly unforgivable sin.